src/pip/_internal/models/release_control.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+
+from pip._vendor.packaging.utils import canonicalize_name
+
+from pip._internal.exceptions import CommandError
+
+
+class ReleaseControl:
+    """Helper for managing which release types can be installed."""
+
+    __slots__ = ["all_releases", "only_final", "_order"]
+
+    def __init__(
+        self,
+        all_releases: set[str] | None = None,
+        only_final: set[str] | None = None,
+    ) -> None:
+        if all_releases is None:
+            all_releases = set()
+        if only_final is None:
+            only_final = set()
+
+        self.all_releases = all_releases
+        self.only_final = only_final
+        # Track the order of arguments as (attribute_name, value) tuples
+        # This is used to reconstruct arguments in the correct order for subprocesses
+        self._order: list[tuple[str, str]] = []
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+
+        # Only compare all_releases and only_final, not _order
+        # The _order list is for internal tracking and reconstruction
+        return (
+            self.all_releases == other.all_releases
+            and self.only_final == other.only_final
+        )
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.all_releases}, {self.only_final})"
+
+    def handle_mutual_excludes(
+        self, value: str, target: set[str], other: set[str], attr_name: str
+    ) -> None:
```

Comment:
Please add a docstring describing what this does. It's not too bad to grok manually, but some help can't hurt :)